### PR TITLE
Preparation for deprecating reactions' user metadata

### PR DESCRIPTION
**Pull request by Niloth-p**
### What does this PR do, and why?
- Prepares for the deprecation of user metadata from messages' reactions and reaction events
    - Supports both `user_id` (ZFL>=2) and `user["id"]` for messages' reactions (with added test cases)
    - Supports both `user_id` (ZFL>=2) and `user["user_id"]` for reaction events (with added test cases)
    - Introduces a common `get_user_id_from_reaction()` for handling the different possible fields
    - Switches to using `_all_users_by_id["user_id"]` to remove dependency on `reaction["user"]["full_name"]
    - Improves the `ReactionEvent` schema to support both versions
- Fixes a bug in reaction event handling (remove op), and adds tests covering the missing cases
- Refactors the reactions' functions a bit before the above changes

### External discussion & connections

- [x] Discussed in **#zulip-terminal** in `reaction metadata`
- [ ] Fully fixes #
- [ ] Partially fixes issue #
- [ ] Builds upon previous unmerged work in PR #
- [ ] Is a follow-up to work in PR #
- [ ] Requires merge of PR #
- [ ] Merge will enable work on #

### How did you test this?

- [x] Manually - Behavioral changes
- [ ] Manually - Visual changes
- [x] Adapting existing automated tests
- [x] Adding automated tests for new behavior (or missing tests)
- [ ] Existing automated tests should already cover this (*only a refactor of tested code*)

### Self-review checklist for each commit
- [x] It is a [minimal coherent idea](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development)
- [x] It has a commit summary following the [documented style](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development) (title & body)
- [x] It has a commit summary describing the  motivation and reasoning for the change
- [x] It individually passes linting and tests
- [x] It contains test additions for any new behavior
- [x] It flows clearly from a previous branch commit, and/or prepares for the next commit

### CODE REVIEW

**Review by mounilKshah — COMMENTED** [review]
Thanks for working on this @Niloth-p 
At first glance, the `get_user_id_from_reaction()` seemed unnecessary to me but on closer look, multiple usages of the same logic make the function a necessity. The coverage of cases is more vast than the impact that I had expected. Overall, this worked as expected across all the use cases that I tried for. 
I think the commit structure can be changed. You can squash the initial commits which work towards simplifying/switching to the new API response. Then the next commit can be the patch providing the new function of `get_user_id_from_reaction()` and its test function. Then you may change all implementations of the older response and add the function calls wherever necessary.
Apart from this, I don't have any other changes to suggest for this PR.

File: zulipterminal/model.py [hunk 1]
```diff
@@ -1849,9 +1849,20 @@ def _handle_reaction_event(self, event: Event) -> None:
...
+                        event["user"]["user_id"]
+                        if event.get("user") and isinstance(event["user"], dict)
+                        else event["user_id"]
+                    )
+                    if (
+                        reaction["emoji_code"] == event["emoji_code"]
+                        and reaction_user_id == event_user_id
+                    ):
                         message["reactions"].remove(reaction)
                         break
```
Comment (timabbott) [inline zulipterminal/model.py:1874]: Feature level 3.0 is old enough that we could drop support for it. But I guess this is OK.

**Comment by timabbott** [comment]
I did a fairly careful read, and this seems like a good PR to me, and I think it makes sense to merge and fix-forward from here as needed. It's sadly a bit of wasted effort -- we could just process `user_id` given no other clients support older versions than 5.0 -- but doing that can be left for future work, as it's not a problem for this code path to support ancient versions, just extra code we don't need.

**Comment by Niloth-p** [comment]
Thank you!
True, I did consider dropping the older version support, but during our GSoC calls, Neil has explicitly expressed how he prefers retaining support for the older versions unlike all the other clients, so this was done for his satisfaction. We've had a few discussions on adding a uniform framework for version-specific features / ZFL-specific handling, so knowing that, I didn't want to be the person who drops the support for older versions.

**Pull request merged**